### PR TITLE
Update the flags of the testsuite

### LINKED ISSUE #599: Separate unit tests from integration tests
Running `pytest` with no arguments currently downloads a tarball from ECMWF's S3, unpacks it, computes standardization stats, builds graphs, and runs a full Lightning training loop. That's a lot to ask for a quick local check.

Beyond the slow feedback loop, mixing test types in a flat directory makes it harder to understand what each test is actually verifying. A reader can't tell at a glance whether a test is checking isolated logic or exercising the full stack, which makes the suite harder to maintain and extend.

Most tests call `init_datastore_example()` from `conftest.py`, which triggers the S3 download. Only a handful are actually fast and isolated:

- **Unit tests** (no I/O): `test_imports.py`, `test_config.py`, `test_cli.py`, `test_time_slicing.py`, and the `test_all_gather_cat_*` functions in `test_training.py`
- **Integration tests** (need real data): `test_datastores.py`, `test_graph_creation.py`, `test_training.py::test_training`, `test_datasets.py`, `test_plotting.py`, `test_plot_graph.py`

**Options**

1. Register a marker in `pyproject.toml`:

```toml
[tool.pytest.ini_options]
markers = [
    "integration: tests that require network access or real data (deselect with '-m not integration')",
]
```

2. Split into `tests/unit/` and `tests/integration/`, moving `conftest.py` (and the S3 fixture) into the integration folder. Cleaner long-term, makes the distinction visible in the file tree, and prevents the fixture from being accidentally used in unit tests.

**Bonus**

`test_clamping.py` could become a true unit test by swapping the real `MDPDatastore` for the `DummyDatastore` that already exists in `tests/dummy_datastore.py`. Same logic, no S3 dependency. `test_time_slicing.py` already does this well and is a good model to follow.

### PR DESCRIPTION
**Pull request by KumarShivam1908**
## Describe your changes

Running the tests took too long, and there was no easy way to skip the slow parts. This PR adds labels (pytest markers) to the tests, so you can choose what to run. Plain `pytest` still runs everything.

### The labels

| Label | What it means | Tests | Who adds it |
|---|---|---|---|
| `requires_real_data` | The test needs the real example weather data (`mdp`, `npyfilesmeps`), which is downloaded the first time | 91 | You, by hand. Tests that loop over `datastore_name` get it automatically |
| `requires_training` | The test trains a model (calls `trainer.fit()`) | 5 | You, by hand |
| `quick` | The test needs neither of the above | 193 | Nobody. It is added automatically, so it is always right |

- The old `slow` label is removed. It was only on the training tests, and `requires_training` now says what they do.
- `test_test_step_excludes_boundary_from_spatial_loss` only runs `trainer.test()`, not training, so it is in `quick`.
- 3 tests have both "requires" labels, because they train on real data: `test_training[mdp]`, `test_training[npyfilesmeps]` and `test_training_output_std`.
- If a test uses the real data but has no `requires_real_data` label, it fails with a message saying to add the label. This keeps the labels correct when new tests are added.

### Which command to run

| Command | What it runs | Tests |
|---|---|---|
| `pytest` | Everything | 286 |
| `pytest -m quick` | Only the fast tests: no real data, no training | 193 |
| `pytest -m "not requires_real_data"` | Everything that works without the real data, including training on fake data | 195 |
| `pytest -m requires_training` | Only the training tests | 5 |

### What changed

- `pyproject.toml`: add the 3 labels, remove `slow`.
- `tests/conftest.py`: add `requires_real_data` and `quick` automatically, and fail tests that use real data without the label.
- Add `requires_real_data` to the tests that always use the real data, and `requires_training` to the 3 tests that call `trainer.fit()`.
- `tests/test_plotting.py`: the plotting tests used to build the same small graph again for every test. Now it is built once and copied. The graph only depends on the grid, which is the same for every `DummyDatastore`.
- `CONTRIBUTING.md`: explain `pytest -m quick` and the labels.

**Why:** #599. `pytest` always downloaded and read the example data, even for a small local check, and nothing showed which tests needed it. As agreed with @sadamov, the labels say what a test needs, `slow` is replaced, and there are no `unit` / `integration` labels.

### Timings

Measured on my machine (Windows, CPU, example data already downloaded):

| Command | Before | After |
|---|---|---|
| `pytest -m quick` (same tests as "not real data and not training") | 70.7 s | 52.0 s |
| `pytest` | 4:09 | 3:53 |

Most of the quick-run saving comes from building the graph once.

### How I checked it

- All 286 tests pass.
- The 91 tests with `requires_real_data` are exactly the tests that create an `mdp` or `npyfilesmeps` datastore. I checked this with a temporary script, not committed.
- `pytest -m quick` runs exactly the same 193 tests as `pytest -m "not requires_real_data and not requires_training"`.
- Every test that calls `trainer.fit()` has `requires_training`. Tests that only run `trainer.test()`, save a checkpoint, or do a single loss step do not.

### Not in this PR

- CI is unchanged. #606 and #577 already edit `install-and-test.yml`, and caching the example data on CI is #257.
- The slowest `quick` test is `test_plot_examples_gif_integration` (about 9 s). It could be made lighter in a follow-up.

**Dependencies:** none. Two open PRs edit the same marker list in `pyproject.toml`, so whichever merges second needs a one-line fix:
- #635 changes the text of the `slow` marker, which this PR removes.
- #728 adds its own `requires_real_data` line.

## Issue Link

closes #599

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

None of the above: this is maintenance of the test setup, plus a short `CONTRIBUTING.md` addition. No library code changes.

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes (not needed: the commands are documented in `CONTRIBUTING.md`)
- [ ] I have added tests that prove my fix is effective or that my feature works (no test of the check itself is committed; it was checked as described above)
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by sadamov** [comment]
i think you forgot to push @KumarShivam1908 ?

**Comment by KumarShivam1908** [comment]
> i think you forgot to push @KumarShivam1908 ?

yeah mb , please check now

File: tests/conftest.py [hunk 1]
```diff
@@ -107,6 +107,63 @@ def download_meps_example_reduced_dataset():
...
+    """Return an ``__init__`` that fails the test constructing ``kind``."""
+
+    def __init__(self, *args, **kwargs):
+        pytest.fail(
+            f"This test constructs the '{kind}' example datastore, which "
+            "needs real example data. Mark it with "
+            "@pytest.mark.requires_real_data."
+        )
+
+    return __init__
```
Comment (sadamov) [inline tests/conftest.py]: For `npyfilesmeps` the guard fires too late: `init_datastore_example` calls `download_meps_example_reduced_dataset()` before `init_datastore`, so an unmarked test still downloads and unpacks the 50 MB tarball and computes the stats.    
```suggestion
def _fail_unmarked_example_datastore(kind):
    """Return a callable that fails the test constructing ``kind``."""

    def fail(*args, **kwargs):
        pytest.fail(
            f"This test constructs the '{kind}' example datastore, which "
            "needs real example data. Mark it with "
            "@pytest.mark.requires_real_data."
        )

    return fail
```

File: tests/conftest.py [hunk 2]
```diff
@@ -107,6 +107,63 @@ def download_meps_example_reduced_dataset():
...
+        return (yield)
+
+    with pytest.MonkeyPatch.context() as mp:
+        for kind in REAL_DATA_DATASTORE_KINDS:
+            mp.setattr(
+                DATASTORES[kind],
+                "__init__",
+                _fail_unmarked_example_datastore(kind),
+            )
+        return (yield)
```
Comment (sadamov) [inline tests/conftest.py:171]:
```suggestion
    with pytest.MonkeyPatch.context() as mp:
        for kind in REAL_DATA_DATASTORE_KINDS:
            mp.setattr(
                DATASTORES[kind],
                "__init__",
                _fail_unmarked_example_datastore(kind),
            )
        # the npyfilesmeps example is downloaded before its datastore is built
        mp.setitem(
            globals(),
            "download_meps_example_reduced_dataset",
            _fail_unmarked_example_datastore("npyfilesmeps"),
        )
        return (yield)
```

Verified locally with a probe: unmarked `npyfilesmeps` and `mdp` tests fail in 3 s with the marker message and no download, a marked `mdp` test passes.

File: CHANGELOG.md [hunk 3]
```diff
@@ -103,6 +103,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Maintenance
 
+- Add `requires_real_data` and `requires_training` pytest markers (the latter replaces `slow`), an automatic `quick` marker for tests needing neither, and build the dummy graph once in the plotting tests, so `pytest -m quick` gives a fast check. [\#750](https://github.com/mllam/neural-lam/pull/750) @KumarShivam1908
```
Comment (sadamov) [inline CHANGELOG.md]: Append to your entry: `; import mlflow lazily in CustomMLFlowLogger so plain import neural_lam no longer loads it.`

**Review by sadamov — APPROVED** [review]
Thanks @KumarShivam1908, looks good. I checked test time locally on my system:

| Command | Tests | pytest time | Wall time | Slowest test |
|---|---:|---:|---:|---|
| `pytest -m quick` | 193 | 42 s | 51 s | `test_plot_examples_gif_integration[1-hours]` 8 s |
| `pytest -m "requires_training and not requires_real_data"` | 2 | 27 s | 36 s | `test_training[dummydata]` 21 s |
| `pytest -m "not requires_real_data"` | 195 | 62 s | 71 s | `test_training[dummydata]` 22 s |
| `pytest -m requires_real_data` | 91 | 168 s | 176 s | `test_training[npyfilesmeps]` 45 s |
| `pytest` (sum of the three disjoint groups) | 286 | 237 s | 263 s | `test_training[npyfilesmeps]` 45 s |

193 quick instead of your 188 because main gained tests since.

One minor point below on the `npyfilesmeps` guard, and one small addition while we are at test speed: `import neural_lam` loads mlflow (457 of 5831 modules) because `custom_loggers.py` imports it at module level and `neural_lam/__init__.py` pulls everything in. Making that import lazy (diff below, the file is not in the PR so I cannot attach suggestion blocks to it) drops the module count by 14% and about 20 s off a cold `pytest -m quick` on our cluster (warm unchanged). Could we include it here, together with a short clause in the CHANGELOG line? Otherwise good to go.

```diff
diff --git a/neural_lam/custom_loggers.py b/neural_lam/custom_loggers.py
index 1d5ff9f3..8b6dd828 100644
--- a/neural_lam/custom_loggers.py
+++ b/neural_lam/custom_loggers.py
@@ -5,8 +5,6 @@ import os
 
 # Third-party
 import matplotlib.pyplot as plt
-import mlflow
-import mlflow.pytorch
 import pytorch_lightning as pl
 from loguru import logger
 
@@ -52,6 +50,10 @@ class CustomMLFlowLogger(pl.loggers.MLFlowLogger):
         self._save_dir = save_dir
         os.makedirs(self._save_dir, exist_ok=True)
 
+        # mlflow is heavy and only needed with --logger mlflow
+        # Third-party
+        import mlflow
+
         mlflow.start_run(run_id=self.run_id, log_system_metrics=True)
         mlflow.set_tag("mlflow.runName", run_name)
         mlflow.log_param("run_id", self.run_id)
@@ -99,6 +101,7 @@ class CustomMLFlowLogger(pl.loggers.MLFlowLogger):
             If AWS credentials for the MLflow artifact store are missing.
         """
         # Third-party
+        import mlflow
         from botocore.exceptions import NoCredentialsError
         from PIL import Image
```

`mlflow.pytorch` was never referenced, so it just goes. `test_custom_loggers.py` keeps working since `@patch("mlflow.log_image")` is resolved at call time.
